### DISCUSSION TITLE
Separation of constraints and objectives

### DISCUSSION THREAD

**Discussion by oneSapling** [TimefoldAI/timefold-solver#1173] [General]
In timefold-solver and optaplanner, it seems that constraints and objectives are processed together. Will you consider defining and processing constraints and objectives separately in the future? If all constraints are merged together for weighted solving, it will be difficult to define the weights between various constraints when there are many constraints.

**Comment by triceo** [comment]
Hello @oneSapling, and thanks for your question.

Although I agree that balancing the weights of constraints is a difficult job, I do not see how introducing another concept would help that. In the end, even with absolute separation between the two categories, you still need to define how hard or soft each constraint and objective is - introducing this separation, in my opinion, does not solve anything.

If you absolutely need such separation, you may use `BendableScore` - this allows you to have an unlimited amount of score levels, each independent of the others. Your constraints can then be put on the lower (more important) levels, and your objectives can go to the higher (less important) levels. The issue of "how important is this constraint relative to the others" does not really go away though - it will never go away, regardless of how much separation there is.
